File: yu-ai-router-go-sdk/http_client.go
package yuairoutersdk

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

const chatCompletionsPath = "/v1/chat/completions"

type httpClient struct {
	cfg    Config
	client *http.Client
}

func newHTTPClient(cfg Config) *httpClient {
	if cfg.HTTPClient != nil {
		return &httpClient{cfg: cfg, client: cfg.HTTPClient}
	}
	return &httpClient{
		cfg: cfg,
		client: &http.Client{
			Timeout: cfg.ConnectTimeout + cfg.ReadTimeout,
		},
	}
}

func (c *httpClient) chat(ctx context.Context, request ChatRequest) (*ChatResponse, error) {
	stream := false
	request.Stream = &stream

	var lastErr error
	maxAttempt := c.cfg.MaxRetries + 1
	for attempt := 1; attempt <= maxAttempt; attempt++ {
		response, err := c.doChatRequest(ctx, request)
		if err == nil {
			return response, nil
		}
		lastErr = err

		var authErr *AuthError
		var limitErr *RateLimitError
		if errors.As(err, &authErr) || errors.As(err, &limitErr) {
			return nil, err
		}
		if attempt < maxAttempt && c.cfg.RetryDelay > 0 {
			time.Sleep(c.cfg.RetryDelay * time.Duration(attempt))
		}
	}
	return nil, lastErr
}

func (c *httpClient) doChatRequest(ctx context.Context, request ChatRequest) (*ChatResponse, error) {
	body, err := json.Marshal(request)
	if err != nil {
		return nil, err
	}

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+chatCompletionsPath, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	httpReq.Header.Set("Authorization", "Bearer "+c.cfg.APIKey)
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, mapHTTPError(resp.StatusCode, string(respBody))
	}
	if apiErr := parseAPIError(respBody); apiErr != nil {
		return nil, apiErr
	}

	var chatResp ChatResponse
	if err = json.Unmarshal(respBody, &chatResp); err != nil {
		return nil, fmt.Errorf("parse chat response failed: %w", err)
	}
	if len(chatResp.Choices) == 0 {
		return nil, fmt.Errorf("invalid chat response: %s", string(respBody))
	}
	return &chatResp, nil
}

func (c *httpClient) chatStream(ctx context.Context, request ChatRequest, callback StreamCallback) error {
	stream := true
	request.Stream = &stream

	body, err := json.Marshal(request)
	if err != nil {
		return err
	}
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+chatCompletionsPath, bytes.NewReader(body))
	if err != nil {
		return err
	}
	httpReq.Header.Set("Authorization", "Bearer "+c.cfg.APIKey)
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(httpReq)
	if err != nil {
		callback.OnError(err)
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		respBody, _ := io.ReadAll(resp.Body)
		err = mapHTTPError(resp.StatusCode, string(respBody))
		callback.OnError(err)
		return err
	}

	reader := bufio.NewReader(resp.Body)
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		line, readErr := reader.ReadString('\n')
		if readErr != nil {
			if readErr == io.EOF {
				callback.OnComplete()
				return nil
			}
			callback.OnError(readErr)
			return readErr
		}

		line = strings.TrimSpace(line)
		if line == "" || !strings.HasPrefix(line, "data:") {
			continue
		}
		data := strings.TrimSpace(strings.TrimPrefix(line, "data:"))
		if data == "" {
			continue
		}

		if data == "[DONE]" {
			callback.OnComplete()
			return nil
		}

		if apiErr := parseAPIError([]byte(data)); apiErr != nil {
			callback.OnError(apiErr)
			return apiErr
		}

		var streamResp StreamResponse
		if err = json.Unmarshal([]byte(data), &streamResp); err != nil {
			// 忽略非 JSON 行，保持与 Java SDK 一致的容错策略
			continue
		}
		if len(streamResp.Choices) == 0 {
			continue
		}
		choice := streamResp.Choices[0]
		if choice.FinishReason == "stop" {
			callback.OnComplete()
			return nil
		}
		chunk := ChatChunk{
			Content:          choice.Delta.Content,
			ReasoningContent: choice.Delta.ReasoningContent,
			Model:            streamResp.Model,
			Done:             false,
		}
		if chunk.Content == "" && chunk.ReasoningContent == "" {
			continue
		}
		callback.OnMessage(chunk)
	}
}

type apiErrorEnvelope struct {
	Code    int             `json:"code"`
	Data    json.RawMessage `json:"data"`
	Message string          `json:"message"`
}

func parseAPIError(raw []byte) error {
	var envelope apiErrorEnvelope
	if err := json.Unmarshal(raw, &envelope); err != nil {
		return nil
	}
	// 避免误判普通 chat 响应，必须同时具备 code 和 message 字段语义
	if envelope.Code == 0 {
		return nil
	}
	message := strings.TrimSpace(envelope.Message)
	if message == "" {
		message = "request failed"
	}
	switch envelope.Code {
	case 40100, 40101:
		return &AuthError{Message: message}
	case 42900:
		return &RateLimitError{Message: message}
	default:
		return &APIError{Code: envelope.Code, Message: message}
	}
}

func mapHTTPError(statusCode int, message string) error {
	message = strings.TrimSpace(message)
	if message == "" {
		message = fmt.Sprintf("request failed with status %d", statusCode)
	}
	switch statusCode {
	case http.StatusUnauthorized:
		return &AuthError{Message: message}
	case http.StatusTooManyRequests:
		return &RateLimitError{Message: message}
	default:
		return &APIError{Code: statusCode, Message: message}
	}
}

File: yu-ai-router-python-sdk/yu_ai_router_sdk/http_client.py
"""HTTP client wrapper for SDK."""

from __future__ import annotations

import json
import time

import httpx

from yu_ai_router_sdk.callback import StreamCallback
from yu_ai_router_sdk.config import ClientConfig
from yu_ai_router_sdk.exceptions import AuthException, RateLimitException, YuAIException
from yu_ai_router_sdk.models import ChatChunk, ChatRequest, ChatResponse, StreamResponse


class HttpClient:
    def __init__(self, config: ClientConfig) -> None:
        self.config = config
        timeout = httpx.Timeout(
            connect=config.connect_timeout / 1000,
            read=config.read_timeout / 1000,
            write=config.write_timeout / 1000,
            pool=config.connect_timeout / 1000,
        )
        self.client = httpx.Client(timeout=timeout)

    def chat(self, request: ChatRequest) -> ChatResponse:
        request.stream = False
        payload = request.to_payload()
        url = f"{self.config.base_url}/v1/chat/completions"
        last_error: Exception | None = None

        for attempt in range(self.config.max_retries + 1):
            try:
                response = self.client.post(url, json=payload, headers=self._headers())
                self._raise_for_status(response)
                return ChatResponse.from_dict(response.json())
            except (AuthException, RateLimitException):
                raise
            except Exception as exc:  # noqa: PERF203
                last_error = exc
                if attempt < self.config.max_retries:
                    time.sleep((self.config.retry_delay * (attempt + 1)) / 1000)

        raise YuAIException(
            f"Request failed after {self.config.max_retries} retries: {last_error}",
        )

    def chat_stream(self, request: ChatRequest, callback: StreamCallback) -> None:
        request.stream = True
        payload = request.to_payload()
        url = f"{self.config.base_url}/v1/chat/completions"

        try:
            with self.client.stream("POST", url, json=payload, headers=self._headers()) as response:
                self._raise_for_status(response)
                completed = False
                for line in response.iter_lines():
                    if not line:
                        continue
                    raw = line.strip()
                    if not raw.startswith("data:"):
                        continue
                    data = raw[5:].strip()
                    if not data:
                        continue
                    try:
                        parsed = json.loads(data)
                        stream_response = StreamResponse.from_dict(parsed)
                    except Exception:
                        continue
                    if not stream_response.choices:
                        continue
                    choice = stream_response.choices[0]
                    if choice.finish_reason == "stop":
                        callback.on_complete()
                        completed = True
                        break
                    if choice.delta.content is None and choice.delta.reasoning_content is None:
                        continue
                    callback.on_message(
                        ChatChunk(
                            content=choice.delta.content,
                            reasoning_content=choice.delta.reasoning_content,
                            done=False,
                            model=stream_response.model,
                        )
                    )
                if not completed:
                    callback.on_complete()
        except Exception as exc:
            callback.on_error(exc)

    def close(self) -> None:
        self.client.close()

    def _headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self.config.api_key}",
            "Content-Type": "application/json",
        }

    @staticmethod
    def _raise_for_status(response: httpx.Response) -> None:
        if response.status_code < 400:
            return
        message = response.text or f"Request failed with code: {response.status_code}"
        if response.status_code == 401:
            raise AuthException(message, status_code=401)
        if response.status_code == 429:
            raise RateLimitException(message, status_code=429)
        raise YuAIException(message, status_code=response.status_code)
